pybossa/sched.py

```diff
@@ -300,16 +347,105 @@ def template_get_locked_task(project_id, user_id=None, user_ip=None,
             timeout = timeout or TIMEOUT
             remaining = float('inf') if calibration else n_answers - taskcount
             if acquire_lock(task_id, user_id, remaining, timeout):
+                # reserve tasks
+                acquire_reserve_task_lock(project_id, task_id, user_id, timeout)
                 return _lock_task_for_user(task_id, project_id, user_id, timeout, calibration)
-
         return []
 
     return template_get_locked_task
 
 
+
+def reserve_task_sql_filters(project_id, reserve_task_keys, exclude):
+    # build sql query filter from task category cache key
+    # return sql filter for matching task category keys and list of
+    # task category keys that qualifies for a given project_id
+
+    filters, category_keys = "", []
+
+    if not (project_id and len(reserve_task_keys)):
+        return filters, category_keys
+
+    # convert task category redis cache key to sql query
+    # ex "co:name:IBM:ticker:IBM_US" would be converted to
+    # "task.info->>'co_name' IN ('IBM')"
+    filters_dict = {}
+    for item in reserve_task_keys:
+        regex_key = "reserve_task:project:{}:category:(.+?):user".format(project_id)
+        data = re.search(regex_key, item)
+        if not data:
+            continue
+
+        category_keys += [item]
+        category = data.group(1)
+        category_fv = category.split(":")
+        for i in range(0, len(category_fv), 2):
+            key, value = category_fv[i], category_fv[i + 1]
+            if key in filters_dict:
+                if value not in filters_dict[key]:
+                    filters_dict[key] += [value]
+            else:
+                filters_dict[key] = [value]
+
+    # TODO: pull task # from category keys, look for values from task._add_user_info
```

Comment:
is there a ticket for this "TODO"?

Reply:
yes, I'm going to create follow-up tickets for this and remaining work. Thanks!